runtimevar/runtimevar.go

```diff
@@ -107,80 +110,196 @@ var (
 	}
 )
 
+type snapshotPlusError struct {
+	snapshot Snapshot
+	err      error
+}
+
 // Variable provides an easy and portable way to watch runtime configuration
 // variables. To create a Variable, use constructors found in provider-specific
 // subpackages.
 type Variable struct {
```

Comment:
I think that `Variable` should be:

```go
type Variable struct {
	dw       driver.Watcher
	provider string // for metric collection

	// For cancelling the background goroutine, and noticing when it has exited.
	backgroundCancel  context.CancelFunc
	backgroundDone <-chan struct{}

	// haveGood is closed when we get the first good value for the variable.
	haveGood chan struct{}

	// A reference to the changed channel at the last time Watch was called.
	lastWatch <-chan struct{}
	// TODO: might need to retain last watch's snapshot data to avoid spurious wakeups if value A -> B -> A,
	// or if the errors disappear quickly.

	mu     sync.Mutex
	changed chan struct{}  // closed when changing any of the other variables and replaced with a new channel
	lastGood Snapshot  // will always be set after haveGood is closed
	currErr error  // only set if the most recent RPC resulted in failure
	closed bool
}
```

Reply:
Done, with (I think) minor changes.

* Added `last` with latest `Snapshot`.
* Renamed `currErr` as `lastErr`.
* Made `lastGood` a pointer so it can be used as a check to see if `haveGood` is already closed or not.
* Moved `lastWatch` under `mu`, as access needs to be protected by `mu`.
* I dropped `closed`, it can more simply be tracked via `lastErr`.

Reply:
Cool! I think this is good. Final thoughts (all optional):

-  You can check for `haveGood` being closed with `select { case <-c.haveGood: /* is closed */ default: /* is not closed */ }`. You don't need the extra pointer.
-  I don't know that `lastWatch` needs to be protected by the mutex. Watch is the only mutator, and it is not supposed to be called from multiple goroutines.

Reply:
#1 -- done.
#2 -- done.

I also fixed a race during shutdown where sometimes `Watch` would return a "context canceled" error from the background goroutine instead of `ErrClosed.